### PR TITLE
Pipes - angela - api muncher

### PR DESCRIPTION
**Pull request by awilson2017**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? | I used Edamam docs to figure out the structure of the API call and then Postman to parse the JSON |
| Describe your API Wrapper.  How did you decide on the methods you created? | The wrapper has methods list the recipes and then find a recipe. These allow for one location that parses the JSON in case data structure changes in the future we can just change it in one place. How I decided, I looked at the Edamam docs and saw there were two applicable queries (q= and r=) and build the methods based on that. In addition I used slack-api as an example. |
| Describe an edge case or failure case test you wrote for your API Wrapper. | I gave the url a bogus API token and raised an ApiError when the request failed.  |
| Explain how VCR aids in testing an API. | It records the JSON data in a yml file. By recording it we can use it instead of making costly API calls for each test run.  |
| What is the Heroku URL of your deployed application? | https://fathomless-hollows-55266.herokuapp.com/

### CODE REVIEW

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+class RecipesController < ApplicationController
+
+  def root
+
+  end
+
+  def index
+    @query = params[:search]
+
+    if params[:search]
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:13]: Because the empty string is truthy in Ruby, even if you search for nothing you won't end up in the error handling here. Instead, you might have something like `if params[:search] && params[:search] != ""`.

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      @recipes = ApiWrapper.list_recipes(params[:search]).paginate(:page => params[:page], :per_page => 10)
+    else
+      flash[:status] = :failure
+      flash[:message] = "Invalid search term, please search for something else"
+    end
+  end
+
+  def new
+
+  end
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:26]: I don't think the `new` action is used in your app.

File: app/controllers/recipes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  def show
+    @recipe = ApiWrapper.find_recipe(params[:id])
+
+    unless @recipe
+      head :not_found
+    end
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:33]: I like this idea, but I do not believe the code will ever get here. Your API wrapper throws an error if the recipe is not found, but here you check for a `nil` response. You have two options to make the connection:

1. Use a `begin`/`rescue` block here to catch the exception
1. Instead of throwing an error, return `nil` from `ApiWrapper.find_recipe` if the recipe can't be found.

File: lib/api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    # https://api.edamam.com/search
+    # https://api.edamam.com/search?q=chicken&app_id=${YOUR_APP_ID}&app_key=${YOUR_APP_KEY}&from=0&to=3&calories=gte%20591,%20lte%20722&health=alcohol-free
+
+    data = HTTParty.get(recipe_link)
+
+
+
+
+    # parse the JSON data in order to get recipe details for the show page
+    if data.empty?
```
Comment (droberts-sea) [inline lib/api_wrapper.rb:44]: It looks like you're missing a call to `check_status` here. If the API returns an error like `not_authed` (as opposed to no results), as is this method won't be able to detect it.

File: test/controllers/recipes_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      #  VCR.use_cassette("find_recipe") do
+      #    proc {
+      #      ApiWrapper.find_recipe("http://www.edamam.com/ontologies/edamam.owl%23recipe_637913ec61d9da69eb451818c3293df2" + "fake-url")}.must_raise ApiWrapper::ApiError
+      #   #  proc {
+      #   #    ApiWrapper.find_recipe(bogus_uri)
+      #   #  }.must_raise ApiWrapper::ApiError
+      #    #
+       #
+      #    get recipe_path params: { search: "chicken", url: "http://www.edamam.com/ontologies/edamam.owl%23recipe_637913ec61d9da69eb451818c3293df2" + "fake-url"}
+      #    must_respond_with :not_found
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:54]: I think this test is failing because the controller action is handling the error incorrectly. See comments above.

File: test/lib/api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        result = ApiWrapper.list_recipes("")
+        result.must_equal []
+      end
+    end
+
+    it "Raises an ApiError when token is bad" do
+      begin # something which might raise an exception
+        origin_token = ApiWrapper::TOKEN
+        ApiWrapper::TOKEN = "bogus_token"
+        VCR.use_cassette("recipes") do
```
Comment (droberts-sea) [inline test/lib/api_wrapper_test.rb:27]: I like this way of testing an invalid token - much simpler than what we did in class.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | yes |
| Errors are reported to the user | some - You've got the `flash` reporting section in `application.html.erb` set up nicely, but I was not able to get it to actually display a message for me. Make sure you're detecting error cases accurately, and actually reporting problems to the user. |
| API Wrapper to handle the API requests | yes |
| Controller testing | Good start, especially around show, but there's still some pieces missing here. See inline. |
| Lib testing | yes |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** |  |
| Link to Trello Board | yes |
| Link to deployed app on Heroku | yes |
|  **Overall** | Good job overall! It seems like you've got a solid grasp on the core learning goals for this assignment.<br><br>One area worth focusing on going forward is error handling, and making sure that errors are plumbed all the way from the source (in this case an API call) to being displayed to a user. There were a few cases where there was a disconnect, and valuable information ended up being lost. Fixing this means knowing whether a method throws an exception, returns a sentinel value like `nil`, or returns something like an empty array, and being able to handle it appropriately wherever the method is called. This sort of information also manifests when coming up with test cases, which is a big portion of why we study testing in this course.<br><br>Other than that I'm pretty happy with this submission. Keep up the hard work! |
